server/service/src/requisition/program_settings/map.rs

```diff
@@ -0,0 +1,93 @@
+        periods,
+        program_suppliers,
+        requisitions_in_period,
+    }: PrepareProgramSettings,
+) -> Vec<ProgramSettings> {
+    settings
+        .into_iter()
+        .map(|program_setting| {
+            // Filter and map (to include available_periods) order type for program setting
+            let order_types = order_types
+                .iter()
+                .filter(|o| {
+                    o.program_requisition_settings_id == program_setting.program_settings_row.id
+                })
+                .map(|order_type| {
+                    // Filter periods for order type, period_is_available
+                    let available_periods = periods
+                        .iter()
+                        .filter(|period| {
+                            period_is_available(
+                                period,
+                                &program_setting,
+                                &order_type,
+                                &requisitions_in_period,
+                            )
+                        })
+                        .map(|p| p.clone())
+                        .collect();
+
+                    // Order type for program settings
+                    OrderType {
+                        available_periods,
+                        order_type: order_type.clone(),
+                    }
+                })
+                .collect();
+
+            // Filter by program_id
+            let suppliers = program_suppliers
+                .iter()
+                .filter(|s| s.program.id == program_setting.program_row.id)
+                .map(|s| s.clone())
+                .collect();
+
+            ProgramSettings {
+                order_types,
+                suppliers,
+                program_requisition_settings: program_setting,
+            }
+        })
+        .collect()
+}
+
+/// Deduce if period is available for order_type based on
+/// matching period_schedule_id and number of requisition that exists for this
+/// order_type and program_id is within order_type.max_order_per_period
+/// note: lowercase match for order type
+fn period_is_available(
+    period: &PeriodRow,
+    setting: &ProgramRequisitionSettings,
+    order_type: &ProgramRequisitionOrderTypeRow,
+    requisitions_in_period: &Vec<RequisitionsInPeriod>,
+) -> bool {
+    if period.period_schedule_id != setting.program_settings_row.period_schedule_id {
+        return false;
+    }
+
+    match requisitions_in_period.iter().find(|requisition_in_period| {
+        requisition_in_period.program_id == setting.program_row.id
+        && requisition_in_period.period_id == period.id
+        // Case insensitive match for order_type
+            && requisition_in_period.order_type.to_lowercase() == order_type.name.to_lowercase()
+    }) {
+        Some(requisitions_in_period) => {
+            requisitions_in_period.count < order_type.max_order_per_period as i64
+        }
+        None => true,
+    }
```

Comment:
nit: I think it would be easier to read if the logic was broken up a little bit more...
```suggestion
    let num_requisitions_in_period = requisitions_in_period
        .iter()
        .filter(|requisition_in_period| {
            requisition_in_period.program_id == setting.program_row.id
                && requisition_in_period.period_id == period.id
                        // Case insensitive match for order_type
            && requisition_in_period.order_type.to_lowercase() == order_type.name.to_lowercase()
        })
        .count();

    // period is available for selection if we have less requisitions than max_order_per_period
    num_requisitions_in_period < order_type.max_order_per_period as usize
```

Reply:
Hmm this is not quite equivalent, requisition_in_period already has a count of how many requisitions are in a period. How does this look d0d613b556f0f4d049543cf79ae6f0d8791a9ec2